compiler: add the `nkNimNodeLit` node kind

zerbina opened the pull request:
## Summary
AST that represents the content of a `NimNode` tree was not detectable
as such without inspecting the surrounding AST. This complicated logic
inspecting the bodies of macros or `.compileTime` procedures, as it had
to specifically ignore the symbol/symbol-choice nodes part of the AST
for `mNBindSym` magic calls.

The existence of `NimNode` literals inside normal AST is now made
explicit and also detectable without further context via a dedicated
node kind, also preventing it from going unnoticed if erroneously
present in places where it shouldn't be.

## Details
- use `nkNimNodeLit` nodes when deserializing `NimNode`s in
  `vmcompilerserdes`
- static `bindSym` calls are now expanded into a `nkNimNodeLit` during
  semantic analysis
- don't use an extra node for providing source information to
  `opBindSym` -- the source information is already provided by the VM
- the arguments to dynamic `bindSym` calls are now considered during
  constant folding
- remove the `opcNBindSym` VM instruction
- remove the now obsolete `rvmInvalidBindSym` report
- propagate error nodes in `semBindSym`
- include `semBindSym` in compiler tracing

## Changes to typed AST
```nim
...
# inside a compile-time evaluation context (macro, `.compileTime` procedure, or `static` expression/statement)

bindSym("s")
# is now represented in the AST as:
# `NimNodeLit(Sym "s")`
# instead of:
# `Call(Sym "bindSym", Sym "s")`

let ident = ...
bindSym(ident)
# is now represented in the AST as:
# `Call(Sym "bindSym", Sym "ident", Sym "brClosed", IntLit ...)
# instead of:
# `Call(Sym "bindSym", Sym "ident", Sym "brClosed", ..., IntLit ...)
```

saem commented on compiler/vm/vm.nim:
```diff
@@ -2584,15 +2584,23 @@ proc rawExecute(c: var TCtx, pc: var int, tos: var StackFrameIndex): RegisterInd
+      #       single ``nnkIdent`` node from a macro, and then passing the
+      #       resulting semantically analysed AST back into another macro as
+      #       a ``typed`` argument.
+      #       But there is also the question if ``semDynamicBindSym`` is
```
We simply should disallow breaking scoping/lookup rules, the more of that we can get rid of the better. Separate PR for that, I'd imagine..

saem commented on compiler/vm/vm.nim:
```diff
@@ -2584,15 +2584,23 @@ proc rawExecute(c: var TCtx, pc: var int, tos: var StackFrameIndex): RegisterInd
+      #       with the information that a symbol lookup should be performed.
+      #       The lookup result is then stored into a register also
+      #       communicated earlier and VM execution is resumed.
+      #       In some way (but not exactly), this is similar to returning a
```
I really like the symmetry, but prefer removing dynamic bind sym

saem commented on compiler/vm/vmgen.nim:
```diff
@@ -551,15 +551,20 @@ template cmpFloatRep(a, b: BiggestFloat): bool =
+    of nkType:          result = a.typ == b.typ
+    of nkStrKinds:      result = a.strVal == b.strVal
+    of nkIntKinds:      result = a.intVal == b.intVal
+    of nkFloatLiterals: result = cmpFloatRep(a.floatVal, b.floatVal)
```
So clean!

saem commented on compiler/vm/vmgen.nim:
```diff
@@ -2610,6 +2604,19 @@ proc gen(c: var TCtx; n: PNode; dest: var TDest; flags: TGenFlags = {}) =
+
+    var tmp: TDest = c.getTemp(n.typ)
+    c.genLit(n, c.toNodeCnst(n[0]), tmp)
+    c.gABC(n, opcNCopyNimTree, dest, tmp)
```
Good

saem reviewed (approved):
I like it, a lot. It's definitely a type of cake that exists at comptime and makes sense to have explicitly.

zerbina committed:
compiler: add the `nkNimNodeLit` node kind

## Summary
AST that represents the content of a `NimNode` tree was not detectable
as such without inspecting the surrounding AST. This complicated logic
inspecting the bodies of macros or `.compileTime` procedures, as it had
to specifically ignore the symbol/symbol-choice nodes part of the AST
for `mNBindSym` magic calls.

The existence of `NimNode` literals inside normal AST is now made
explicit and also detectable without further context via a dedicated
node kind, also preventing it from going unnoticed if erroneously
present in places where it shouldn't be.

## Details
- use `nkNimNodeLit` nodes when deserializing `NimNode`s in
  `vmcompilerserdes`
- static `bindSym` calls are now expanded into a `nkNimNodeLit` during
  semantic analysis
- don't use an extra node for providing source information to
  `opBindSym` -- the source information is already provided by the VM
- the arguments to dynamic `bindSym` calls are now considered during
  constant folding
- remove the `opcNBindSym` VM instruction
- remove the now obsolete `rvmInvalidBindSym` report
- propagate error nodes in `semBindSym`
- include `semBindSym` in compiler tracing

## Changes to typed AST
```nim
...
# inside a compile-time evaluation context (macro, `.compileTime` procedure, or `static` expression/statement)

bindSym("s")
# is now represented in the AST as:
# `NimNodeLit(Sym "s")`
# instead of:
# `Call(Sym "bindSym", Sym "s")`

let ident = ...
bindSym(ident)
# is now represented in the AST as:
# `Call(Sym "bindSym", Sym "ident", Sym "brClosed", IntLit ...)
# instead of:
# `Call(Sym "bindSym", Sym "ident", Sym "brClosed", ..., IntLit ...)
```

zerbina commented:
bors r+
